tests/functional/test_endpoints_v2.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2012-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import json
+import os
+
+import pytest
+
+from botocore.endpoints_v2 import (
+    EndpointProvider,
+    EndpointResolutionError,
+    RuleSetEndpoint,
+)
+from botocore.loaders import Loader
+
+
+@pytest.fixture
+def partitions():
+    loader = Loader()
+    return loader.load_data('partitions')
+
+
+def ruleset_testcases(case_type):
+    loader = Loader()
+    services = loader.list_available_services('service-2')
+    test_data_dir = os.path.join(
+        os.path.dirname(__file__), 'data', 'endpoints'
+    )
+    services = loader.list_available_services('service-2')
+    for service in services:
+        try:
+            ruleset = loader.load_service_model(service, 'endpoint-rule-set')
+            with open(
+                os.path.join(test_data_dir, service, 'endpoint-tests.json')
+            ) as f:
+                data = f.read()
+                tests = json.loads(data)
+                for test in tests['testCases']:
+                    input_params = test['params']
+                    expected_object = test['expect']
+                    if case_type in expected_object:
+                        yield (ruleset, input_params, expected_object)
+        except:
+            continue
+
+
+@pytest.mark.parametrize(
+    "ruleset,input_params,expected_object", ruleset_testcases('error')
+)
+def test_endpoint_resolution_raises(
+    partitions, ruleset, input_params, expected_object
+):
+    endpoint_provider = EndpointProvider(ruleset, partitions)
+    with pytest.raises(EndpointResolutionError) as err:
+        endpoint_provider.resolve_endpoint(**input_params)
+        assert err.msg == expected_object['error']
```

Comment:
Two things about this line:

* You should dedent this line one level, making assertions about the raised exception should happen outside the `with` block where exceptions are captured. That's what [the docs](https://docs.pytest.org/en/7.1.x/how-to/assert.html) do, and I vaguely remember getting myself into some trouble by not doing the same (but can't remember what the nature of the trouble was)
* Once you've done the above, you can use all of pytest's machinery for asserting against exceptions. However, it also means that your `msg` property is now in `err.value.msg` (the exception object becomes wrapped in the `value` property of the `ExceptionInfo` instance when the `raises` context manager exits, [docs](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.ExceptionInfo))

Reply:
got it thank you! i'll update the context manager variable name to be a clearer as well